bayesian_changepoint_detection_experiment/02_changepoint_exploration_experiment.py

```diff
@@ -0,0 +1,209 @@
+    key: Word2Vec.load(sorted(word_model_map[key])[0]) for key in word_model_map
+}
+
+word_freq_count_cutoff = 30
+
+word_model_cutoff_map = {
+    key: {
+        "model": word_model_loaded_map[key],
+        "cutoff_index": min(
+            map(
+                lambda x: 999999
+                if word_model_loaded_map[key].wv.get_vecattr(x[1], "count")
+                > word_freq_count_cutoff
+                else x[0],
+                enumerate(word_model_loaded_map[key].wv.index_to_key),
+            )
+        ),
+    }
+    for key in word_model_loaded_map
+}
+
+
+# # Examine Handpicked Tokens with a high chance of Change
+
+def examine_words_at_timepoint_range(
+    word_model_map: dict, years_to_examine: list, tok: str = "the", topn: int = 25
+):
+    for year in years_to_examine:
+        print(year)
+        display.display(
+            word_model_map[year]["model"].wv.most_similar(
+                tok, topn=topn, clip_end=word_model_map[year]["cutoff_index"]
+            )
+        )
+        print()
+
+
+# ## Pandemic
+
+# COVID-19. The drastic shift that came from the pandemic that was first talked about in 2019 but became more prevalent in 2020.
+
+(
+    p9.ggplot(change_point_df >> ply.query("tok=='pandemic'"))
+    + p9.aes(x="year_pair", y="changepoint_prob", group=0)
+    + p9.geom_point()
+    + p9.geom_line()
+    + p9.coord_flip()
+    + p9.theme_seaborn("white")
+    + p9.labs(
+        x="years",
+        y="Probability of Changepoint",
+        title="Changepoint Prediction for Token ('pandemic')",
+    )
+)
+
+examine_words_at_timepoint_range(
+    word_model_cutoff_map, [2017, 2018, 2019, 2020], tok="pandemic", topn=20
+)
+
+# ## Rituximab
+
+# Not sure the change here but seems like the time period is when the drug was being used to treat rheumatoid arthritis. More investigation is needed.
+
+(
+    p9.ggplot(change_point_df >> ply.query("tok=='rituximab'"))
+    + p9.aes(x="year_pair", y="changepoint_prob", group=0)
+    + p9.geom_point()
+    + p9.geom_line()
+    + p9.coord_flip()
+    + p9.theme_seaborn("white")
+    + p9.labs(
+        x="years",
+        y="Probability of Changepoint",
+        title="Changepoint Prediction for Token ('rituximab')",
+    )
+)
+
+examine_words_at_timepoint_range(
+    word_model_cutoff_map,
+    [2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007],
+    tok="rituximab",
+    topn=20,
+)
+
+# ## Asthma
+
+# There is semantic shift for asthma where in 2002-2003 seems like focus shifted from co-morbidity to drug treatments. Fun fact asthma is connected to type 2 diabetes (had no idea).
+
+(
+    p9.ggplot(change_point_df >> ply.query("tok=='asthmatics'"))
+    + p9.aes(x="year_pair", y="changepoint_prob", group=0)
+    + p9.geom_point()
+    + p9.geom_line()
+    + p9.coord_flip()
+    + p9.theme_seaborn("white")
+    + p9.labs(
+        x="years",
+        y="Probability of Changepoint",
+        title="Changepoint Prediction for Token ('asthmatics')",
+    )
+)
+
```

Comment:
why are there 2 change points? Is it changing from co-morbidity to drug treatments back to co-morbidity?

Reply:
>why are there 2 change points? Is it changing from co-morbidity to drug treatments back to co-morbidity?

Good question. I think whats happening here is one of two things. 
1. The model isn't sure on when the timepoint change occurred. 
2. There were two changepoint shifts for this token. 

I'm inclined to say its the latter based on the concepts from the upset plots (uploading shortly).